src/blueapi/cli/amq.py

```diff
@@ -21,60 +13,49 @@ def __init__(self, message: str) -> None:
 
 class AmqClient:
     app: MessagingTemplate
+    complete: threading.Event
 
     def __init__(self, app: MessagingTemplate) -> None:
         self.app = app
+        self.complete = threading.Event()
+
+    def __enter__(self) -> None:
+        self.app.connect()
+
+    def __exit__(self, exc_type, exc_value, exc_traceback) -> None:
+        self.app.disconnect()
 
-    def run_plan(
+    def subscribe_to_topics(
         self,
-        name: str,
-        params: Mapping[str, Any],
+        task_id: str,
         on_event: Optional[Callable[[WorkerEvent], None]] = None,
-        on_progress_event: Optional[Callable[[ProgressEvent], None]] = None,
-        timeout: Optional[float] = None,
-    ) -> str:
-        complete = threading.Event()
+    ) -> None:
+        """Run callbacks on events/progress events with a given correlation id."""
 
         def on_event_wrapper(ctx: MessageContext, event: WorkerEvent) -> None:
-            if on_event is not None:
+            if (on_event is not None) and (ctx.task_id == task_id):
                 on_event(event)
 
-            if event.is_complete():
-                complete.set()
-                if event.is_error():
-                    raise BlueskyRemoteError(str(event.errors) or "Unknown error")
+            if (event.is_complete()) and (ctx.task_id == task_id):
+                self.complete.set()
 
-        def on_progress_event_wrapper(
-            ctx: MessageContext, event: ProgressEvent
-        ) -> None:
-            if on_progress_event is not None:
-                on_progress_event(event)
-
-        self.app.subscribe(
-            self.app.destinations.topic("public.worker.event"), on_event_wrapper
-        )
         self.app.subscribe(
             self.app.destinations.topic("public.worker.event"),
-            on_progress_event_wrapper,
+            on_event_wrapper,
         )
 
-        # self.app.send("worker.run", {"name": name, "params": params})
-        task_response = self.app.send_and_receive(
-            "worker.run", {"name": name, "params": params}, reply_type=TaskResponse
-        ).result(5.0)
-        task_id = task_response.task_id
-
-        if timeout is not None:
-            complete.wait(timeout)
-
-        return task_id
-
-    def get_plans(self) -> PlanResponse:
-        return self.app.send_and_receive(
-            "worker.plans", PlanRequest(), PlanResponse
-        ).result(5.0)
-
-    def get_devices(self) -> DeviceResponse:
-        return self.app.send_and_receive(
-            "worker.devices", DeviceRequest(), DeviceResponse
-        ).result(5.0)
+    def wait_for_complete(self, timeout: Optional[float] = None) -> None:
+        begin_time = time.time()
+        timedout = False
+        while not self.complete.is_set():
+            current_time = time.time()
+            if (timeout is not None) and ((current_time - begin_time) > timeout):
+                timedout = True
+                break
+
+        if timedout:
```

Comment:
I believe the following will work, but test it:
```suggestion
        self.complete.wait(timeout=timeout)
        
        if not self.complete.is_set:
```